node/cmd/spy/vaa_verifier.go

```diff
@@ -0,0 +1,136 @@
+	return &VaaVerifier{
+		logger:       logger,
+		rpcUrl:       rpcUrl,
+		coreAddr:     ethCommon.HexToAddress(coreAddr),
+		guardianSets: make(map[uint32]*common.GuardianSet),
+	}
+}
+
+// GetInitialGuardianSet gets the current guardian set and adds it to the map. It is not necessary
+// to call this function, but doing so will allow you to verify that the RPC endpoint works on start up,
+// rather than having it fail the first VAA is received.
+func (v *VaaVerifier) GetInitialGuardianSet() error {
+	timeout, cancel := context.WithTimeout(context.Background(), RpcTimeout)
+	defer cancel()
+
+	rawClient, err := ethRpc.DialContext(timeout, v.rpcUrl)
+	if err != nil {
+		return fmt.Errorf("failed to connect to ethereum: %w", err)
+	}
+
+	client := ethClient.NewClient(rawClient)
+	caller, err := ethAbi.NewAbiCaller(v.coreAddr, client)
+	if err != nil {
+		return fmt.Errorf("failed to create caller: %w", err)
+	}
+
+	gsIndex, err := caller.GetCurrentGuardianSetIndex(&ethBind.CallOpts{Context: timeout})
+	if err != nil {
+		return fmt.Errorf("error requesting current guardian set index: %w", err)
+	}
+
+	result, err := caller.GetGuardianSet(&ethBind.CallOpts{Context: timeout}, gsIndex)
+	if err != nil {
+		return fmt.Errorf("error requesting guardian set for index %d: %w", gsIndex, err)
+	}
+
+	gs := &common.GuardianSet{
+		Keys:  result.Keys,
+		Index: gsIndex,
+	}
+
+	v.logger.Warn("read current guardian set", zap.Uint32("index", gsIndex), zap.Any("gs", *gs))
+	v.guardianSets[gsIndex] = gs
+	return nil
+}
+
+// VerifySignatures verifies that the signature on a VAA is valid, based on the guardian set contained in the VAA.
+// If the guardian set is not currently in our map, it queries that guardian set and adds it.
+func (v *VaaVerifier) VerifySignatures(vaaBytes []byte) (bool, error) {
+	vv, err := vaa.Unmarshal(vaaBytes)
+	if err != nil {
+		v.logger.Error("failed to unmarshal VAA", zap.Error(err), zap.Any("vaa", vaaBytes))
+		return false, err
+	}
+
+	v.lock.Lock()
+	defer v.lock.Unlock()
+
+	gs, exists := v.guardianSets[vv.GuardianSetIndex]
+	if !exists {
+		gs, err = v.fetchGuardianSet(vv.GuardianSetIndex)
+		if err != nil {
+			return false, fmt.Errorf("failed to fetch guardian set for index %d: %w", vv.GuardianSetIndex, err)
+		}
+
+		v.logger.Warn("read guardian set", zap.Uint32("index", gs.Index), zap.Any("gs", *gs))
+		v.guardianSets[gs.Index] = gs
+	}
+
+	if err := vv.Verify(gs.Keys); err != nil {
+		return false, nil
+	}
+
+	return true, nil
+}
+
+// fetchGuardianSet reads the guardian set for the index passed in.
+func (v *VaaVerifier) fetchGuardianSet(gsIndex uint32) (*common.GuardianSet, error) {
+	timeout, cancel := context.WithTimeout(context.Background(), RpcTimeout)
+	defer cancel()
+
+	rawClient, err := ethRpc.DialContext(timeout, v.rpcUrl)
+	if err != nil {
+		return nil, fmt.Errorf("failed to connect to ethereum: %w", err)
+	}
+
+	client := ethClient.NewClient(rawClient)
```

Comment:
If this is used very frequently we might want to consider reusing the same client instead of creating a new one on each invocation. But I would assume it to be otherwise since we are only invoking this method when we don't have the guardian set in the map.

Reply:
I would expect this to only ever be called when the guardian set changes while the spy is running. In the history of Wormhole v2, that has only happened 3 times, so I'm not too worried at this stage. Given the infrequency, seems better than keeping the client around.

Reply:
It will be used *very* infrequently, only when the guardian set index changes (once a year??).